File: frontend/src/trust/commands.ts
import type { TrustEverythingRequest } from "../generated/TrustEverythingRequest";
import type { TrustEverythingResponse } from "../generated/TrustEverythingResponse";
import type { TrustListResponse } from "../generated/TrustListResponse";
import type { TrustProbeRequest } from "../generated/TrustProbeRequest";
import type { TrustProbeResponse } from "../generated/TrustProbeResponse";
import type { TrustRevokeRequest } from "../generated/TrustRevokeRequest";
import type { TrustRevokeResponse } from "../generated/TrustRevokeResponse";
import type { TrustSetRequest } from "../generated/TrustSetRequest";
import type { TrustSetResponse } from "../generated/TrustSetResponse";
import type { TrustStatusRequest } from "../generated/TrustStatusRequest";
import type { TrustStatusResponse } from "../generated/TrustStatusResponse";
import type { InvokeOptions, IpcClient } from "../ipc";

/**
 * Typed wrappers for the kernel's `trust.*` commands (Task 1.13, REQ-FS-005).
 */

export const TRUST_COMMANDS = {
  status: "trust.status",
  set: "trust.set",
  revoke: "trust.revoke",
  list: "trust.list",
  setTrustEverything: "trust.setTrustEverything",
  probe: "trust.probe",
} as const;

/** Emitted when trust decisions or the enabled flag change. */
export const TRUST_CHANNEL = "trust:changed";

export function trustStatus(
  client: IpcClient,
  paths: readonly string[],
  options?: InvokeOptions,
): Promise<TrustStatusResponse> {
  const payload: TrustStatusRequest = { paths: [...paths] };
  return client.invoke<TrustStatusRequest, TrustStatusResponse>(
    TRUST_COMMANDS.status,
    payload,
    options,
  );
}

export function trustSet(
  client: IpcClient,
  request: TrustSetRequest,
  options?: InvokeOptions,
): Promise<TrustSetResponse> {
  return client.invoke<TrustSetRequest, TrustSetResponse>(TRUST_COMMANDS.set, request, options);
}

export function trustRevoke(
  client: IpcClient,
  request: TrustRevokeRequest,
  options?: InvokeOptions,
): Promise<TrustRevokeResponse> {
  return client.invoke<TrustRevokeRequest, TrustRevokeResponse>(
    TRUST_COMMANDS.revoke,
    request,
    options,
  );
}

export function trustList(client: IpcClient, options?: InvokeOptions): Promise<TrustListResponse> {
  return client.invoke<Record<string, never>, TrustListResponse>(TRUST_COMMANDS.list, {}, options);
}

export function trustSetEverything(
  client: IpcClient,
  request: TrustEverythingRequest,
  options?: InvokeOptions,
): Promise<TrustEverythingResponse> {
  return client.invoke<TrustEverythingRequest, TrustEverythingResponse>(
    TRUST_COMMANDS.setTrustEverything,
    request,
    options,
  );
}

export function trustProbe(
  client: IpcClient,
  request: TrustProbeRequest,
  options?: InvokeOptions,
): Promise<TrustProbeResponse> {
  return client.invoke<TrustProbeRequest, TrustProbeResponse>(
    TRUST_COMMANDS.probe,
    request,
    options,
  );
}

File: crates/helix-trust/src/commands.rs
//! Request and response payloads for the `trust.*` IPC commands.

use serde::{Deserialize, Serialize};
use ts_rs::TS;

use crate::model::{RootTrustStatus, TrustCapability, TrustDecision, WorkspaceTrustMode};

pub const STATUS: &str = "trust.status";
pub const SET: &str = "trust.set";
pub const REVOKE: &str = "trust.revoke";
pub const LIST: &str = "trust.list";
pub const SET_TRUST_EVERYTHING: &str = "trust.setTrustEverything";
pub const PROBE: &str = "trust.probe";

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
#[serde(default)]
pub struct TrustStatusRequest {
    pub paths: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustStatusResponse {
    pub enabled: bool,
    pub trust_everything: bool,
    pub store_healthy: bool,
    pub workspace_mode: WorkspaceTrustMode,
    pub roots: Vec<RootTrustStatus>,
    pub pending_prompts: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustSetRequest {
    pub path: String,
    pub decision: TrustDecision,
    pub inherit_to_children: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustSetResponse {
    pub applied: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustRevokeRequest {
    pub path: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustRevokeResponse {
    pub revoked: bool,
    pub terminated_processes: u32,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustListResponse {
    pub entries: Vec<TrustedFolderEntry>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustedFolderEntry {
    pub path: String,
    pub inherit_to_children: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustEverythingRequest {
    pub enabled: bool,
    pub acknowledged_warning: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustEverythingResponse {
    pub enabled: bool,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustProbeRequest {
    pub path: String,
    pub capability: TrustCapability,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, TS)]
#[ts(export, export_to = "../../../frontend/src/generated/")]
pub struct TrustProbeResponse {
    pub allowed: bool,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn command_names_are_namespaced_under_trust() {
        for name in [STATUS, SET, REVOKE, LIST, SET_TRUST_EVERYTHING, PROBE] {
            assert!(name.starts_with("trust."), "{name}");
        }
    }
}
